Modules/_abc.c

```diff
@@ -0,0 +1,894 @@
+        PyList_SET_ITEM(gs->pending, pos, Py_None);
+        if (PySet_Discard(gs->data, ref) < 0) {
+            Py_DECREF(ref);
+            return -1;
+        }
+        Py_DECREF(ref);
+    }
+    return PyList_SetSlice(gs->pending,
+                           0, PyList_GET_SIZE(gs->pending), NULL);
+}
+
+PyDoc_STRVAR(_reset_registry_doc,
+"Internal ABC helper to reset registry of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_registry(PyObject *m, PyObject *args)
+{
+    PyObject *self;
+    if (!PyArg_UnpackTuple(args, "_reset_registry", 1, 1, &self)) {
+        return NULL;
+    }
+    _abc_data *impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    if (PySet_Clear(impl->_abc_registry->data) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+PyDoc_STRVAR(_reset_caches_doc,
+"Internal ABC helper to reset both caches of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_caches(PyObject *m, PyObject *args)
+{
+    PyObject *self;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_reset_caches", 1, 1, &self)) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    if (PySet_Clear(impl->_abc_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    /* also the second cache */
+    if (PySet_Clear(impl->_abc_negative_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+PyDoc_STRVAR(_get_dump_doc,
+"Internal ABC helper for cache and registry debugging.\n\
+\n\
+Return shallow copies of registry, of both caches, and\n\
+negative cache version. Don't call this function directly,\n\
+instead use ABC._dump_registry() for a nice repr.");
+
+static PyObject *
+_get_dump(PyObject *m, PyObject *args)
+{
+    PyObject *self, *registry, *cache, *negative_cache;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_get_dump", 1, 1, &self)) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    registry = PyObject_CallMethod(impl->_abc_registry->data, "copy", NULL);
+    if (registry == NULL) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    cache = PyObject_CallMethod(impl->_abc_cache, "copy", NULL);
+    if (cache == NULL) {
+        Py_DECREF(impl);
+        Py_DECREF(registry);
+        return NULL;
+    }
+    negative_cache = PyObject_CallMethod(impl->_abc_negative_cache, "copy", NULL);
+    if (negative_cache == NULL) {
+        Py_DECREF(impl);
+        Py_DECREF(registry);
+        Py_DECREF(cache);
+        return NULL;
+    }
+    Py_INCREF(impl->_abc_negative_cache_version); /* PyTuple_Packdoesn't do this. */
+    PyObject *res = PyTuple_Pack(4,
+            registry, cache, negative_cache, impl->_abc_negative_cache_version);
```

Comment:
What ff `PyTuple_Pack` fails?  You'll need to decref `impl`, `registry`, `cache`, and `impl->_abc_negative_cache_version`.

Reply:
No.   At here, all four reference is borrowed from impl.  We don't have strong reference.
PyTuple_Pack() incref them if it's succeeded.
So no need to decref.

Reply:
Right, thanks, Inada-san.

Reply:
These references are new objects returned by `.copy()`, so if `PyTuple_Pack()` does `INCREF` them, then we need to `DECREF` them always at the end. Is this right?

Reply:
Ah, you're right.  When I touched here, they were borrowed.  But it's not for now.